[Proposal] Support Multiple Prefill + Decode in a loop

kushrast opened the pull request:
We would like to support multi-turn conversations with the AR-N model by allowing prefill + decode to be called in a loop without resetting the internal KV cache state.

Example:
_Initialize Runner and KV Cache_
Prompt: "Call David"
Response: "Okay, Call David Lee?"
Prompt: "No, call David Smith"
Response: "Okay, Calling David Smith, not David Lee"
_Clear KV Cache_

Assumptions:
- The assumption here is that to support multiple prefill + decode in a loop, we need to update the prefill_input_pos and prefill_attention_mask to reflect the previously decoded tokens + new prompt tokens.
- An additional assumption is that we need to update the pointers for k_cache and v_cache for prefill

What this PR does:
- Add a function `update_kv_to_prefill_io` to advance prefill pointers for v_cache. Also sets attention_mask up to pos (to cover tokens generated during decode).
- Update fill_prefill_toks to take in previous tokens prefilled + generated so far and using this to set input_pos and attention mask
- Updates test runner to save number of tokens generated and pass to IO Manager. Also comments out resetting KV Cache state so it gets re-used.

Current State:
The code does not crash, but also does not produce proper input. Tested on Samsung S24 with QNN 2.28 binaries

`./qnn_llama3_2_runner --model_path hybrid_llama_qnn.pte    --tokenizer_path tiktokenizer.bin  --eval_mode 1 --prompt "Call David" --kv_updater "ShiftPointer" --logits_scale 0.1 --output_path output.txt --num_iters 2`

Example with a model trained for communication:
<|begin_of_text|><|start_header_id|>user<|end_header_id|>

Call David<|eot_id|><|start_header_id|>assistant<|end_header_id|>

Okay, call David
<|begin_of_text|><|start_header_id|>user<|end_header_id|>

Call David<|eot_id|><|start_header_id|>assistant<|end_header_id|>

renamerenamerenamehabihabihabihabihabihabihabihabi date date date date date date握 culturesogh MMI MMI MMIhabihabihabihabihabislidesванetus Tangourt Abrams datefeedingfeedinghabi Date dateolved MMIhabihabihabihabihabihabihabihabihabihabihabihabi族OGLEalse date date date dateucusucusucushabihabi

sxu commented:
> but we are unsure if we need to update the pointers for the k_cache.

Yeah I think each cache input pointer need to be updated to base address + input_pos.

sxu commented:
> > but we are unsure if we need to update the pointers for the k_cache.
> 
> Yeah I think each cache input pointer need to be updated to base address + input_pos.

Just to elaborate, for shift pointers when switching between one method to another, the followings are needed:
1) update K cache (need to scatter the update because K is transposed) and V cache content (simply memcpy).
2) prepare new method's KV input cache pointers: each K cache points to base address + input_pos (again, because K is transposed), each V cache points to base address + input_pos * head_dim.
3) update new methods mask.

The update from prefill to kv_io is already implemented by `update_prefill_to_kv_io`, it performs all of 1~3 outlined above. The new  `update_kv_to_prefill_io` is only doing 2) for V caches. I think you at least need to set the points for K caches as well. You could also add 1) and 3), but they can also be performed by calling other existing functions (`update_kv_io` for 1, `fill_prefill_toks` for 3), it's up to you.

kushrast commented:
> > > but we are unsure if we need to update the pointers for the k_cache.
> > 
> > 
> > Yeah I think each cache input pointer need to be updated to base address + input_pos.
> 
> Just to elaborate, for shift pointers when switching between one method to another, the followings are needed:
> 
> 1. update K cache (need to scatter the update because K is transposed) and V cache content (simply memcpy).
> 2. prepare new method's KV input cache pointers: each K cache points to base address + input_pos (again, because K is transposed), each V cache points to base address + input_pos * head_dim.
> 3. update new methods mask.
> 
> The update from prefill to kv_io is already implemented by `update_prefill_to_kv_io`, it performs all of 1~3 outlined above. The new `update_kv_to_prefill_io` is only doing 2) for V caches. I think you at least need to set the points for K caches as well. You could also add 1) and 3), but they can also be performed by calling other existing functions (`update_kv_io` for 1, `fill_prefill_toks` for 3), it's up to you.

Thanks for the feedback. I updated `update_kv_to_prefill_io` to do steps 2 and 3. I am assuming 1 is done by the last call to `update_kv_io` in `kv_execute`.

haowhsu-quic commented on examples/qualcomm/oss_scripts/llama/runner/io_manager.cpp:
```diff
@@ -494,6 +494,45 @@ void ShiftPointerIoMgr::prepare_prefill_io(
+      v_cache_in_[prefill_forward_name_];
+  std::vector<std::unique_ptr<executorch::aten::TensorImpl>>& v_cache_out =
+      v_cache_out_[prefill_forward_name_];
+  for (int i = 0, v_cache_stride = head_dim_ * pos; i < v_cache_in.size();
```
Since the pointers of `v_cache_in/out` might have been updated few rounds (ARN + KV + ARN + KV...) before in your scenario, the base pointers will not stay in the initial state.
So if the `pos` is an absolute value, the pointers are going to be put beyond the expected positions. (same issues to `k_cache`)

haowhsu-quic commented on examples/qualcomm/oss_scripts/llama/runner/io_manager.cpp:
```diff
@@ -494,6 +494,45 @@ void ShiftPointerIoMgr::prepare_prefill_io(
+    k_cache_out[i]->set_data(
+        k_cache_out[i]->mutable_data<uint8_t>() + k_cache_stride);
+  }
+
```
Probably need to do a deep copy for `k_cache_in` because the consumed tensor size is different between prefill (prefill_cache_len_ * head_dim_) & decode (kv_cache_len_ * head_dim_) mode:
```cpp
size_t copied_size = pos * sizeof(uint8_t);
for (int j = head_dim_; j > -1; --j) {
  // assume current pointer has been shifted to absolute position of `pos`
  uint8_t* ptr_in = k_cache_in[i]->mutable_data<uint8_t>() - pos;
  memcpy(
    ptr_in + j * prefill_cache_len_,
    ptr_in + j * kv_cache_len_,
    copied_size);
}
```

haowhsu-quic commented on examples/qualcomm/oss_scripts/llama/runner/io_manager.cpp:
```diff
@@ -494,6 +494,48 @@ void ShiftPointerIoMgr::prepare_prefill_io(
+    std::vector<std::vector<executorch::aten::Tensor>>& output_tensors) {
+  // update v_cache
+  assert(pos <= 512);
+  std::vector<std::unique_ptr<executorch::aten::TensorImpl>>& v_cache_in =
```
Could you maintain a member variable `last_pos_` in current `IoMgr`?
When doing updating, we should only shift the difference instead of absolute value: `int64_t pos_diff = pos - last_pos_`.
Use this variable for `k/v` as well:
```c++
for (int i = 0, v_cache_stride = head_dim_ * pos_diff; i < v_cache_in.size(); ++i) {
  v_cache_in[i]->set_data(v_cache_in[i]->mutable_data<uint8_t>() + v_cache_stride);
  v_cache_out[i]->set_data(v_cache_out[i]->mutable_data<uint8_t>() + v_cache_stride);
}
for (int i = 0, k_cache_stride = pos_diff * sizeof(uint8_t); i < k_cache_in_.size(); ++i) {
  k_cache_in[i]->set_data(k_cache_in[i]->mutable_data<uint8_t>() + k_cache_stride);
}
```

haowhsu-quic commented on examples/qualcomm/oss_scripts/llama/runner/io_manager.cpp:
```diff
@@ -494,6 +494,48 @@ void ShiftPointerIoMgr::prepare_prefill_io(
+        k_cache_in[i]->mutable_data<uint8_t>() + k_cache_stride);
+        // assume current pointer has been shifted to absolute position of `pos`
+    uint8_t* ptr_in = k_cache_in[i]->mutable_data<uint8_t>() - pos;
+    for (int j = head_dim_; j > -1; --j) {
```
Please help change to `for (int j = 0; j < head_dim_; ++j)` in order not to override existent `k_cache_in`.

Commit:
Adding KV to Prefill IO (#9466)

Summary: Pull Request resolved: https://github.com/pytorch/executorch/pull/9466

Differential Revision: D71567692

kushrast commented:
@haowhsu-quic I updated the PR with your comments - still seeing bad output but I think we are setting last position correctly.

haowhsu-quic commented on examples/qualcomm/oss_scripts/llama/runner/io_manager.cpp:
```diff
@@ -494,10 +494,56 @@ void ShiftPointerIoMgr::prepare_prefill_io(
+  for (int i = 0, v_cache_stride = head_dim_ * pos_diff; i < v_cache_in.size();
+       i++) {
+    v_cache_in[i]->set_data(
+        v_cache_in[i]->mutable_data<uint8_t>() + v_cache_stride);
```
`v_cache_out` needs to be updated as well, please refer to the resolved comment, thank you.

haowhsu-quic commented on examples/qualcomm/oss_scripts/llama/runner/io_manager.cpp:
```diff
@@ -494,10 +494,56 @@ void ShiftPointerIoMgr::prepare_prefill_io(
+       i++) {
+    k_cache_in[i]->set_data(
+        k_cache_in[i]->mutable_data<uint8_t>() + k_cache_stride);
+    uint8_t* ptr_in = k_cache_in[i]->mutable_data<uint8_t>() - pos_diff;
```
Here we need to get the origin for deep copy: `uint8_t* ptr_in = k_cache_in[i]->mutable_data<uint8_t>() - pos;`

cccclai commented:
@haowhsu-quic do you mean changes like this? 
```
+
+  // update v_cache
+  std::vector<std::unique_ptr<executorch::aten::TensorImpl>>& v_cache_out =
+      v_cache_out_[prefill_forward_name_];
+
+  for (int i = 0, v_cache_stride = head_dim_ * pos_diff; i < v_cache_in.size(); ++i) {
+    v_cache_in[i]->set_data(v_cache_in[i]->mutable_data<uint8_t>() + v_cache_stride);
+    v_cache_out[i]->set_data(v_cache_out[i]->mutable_data<uint8_t>() + v_cache_stride);
   }
 
   // update k_cache
@@ -521,7 +525,7 @@
        i++) {
     k_cache_in[i]->set_data(
         k_cache_in[i]->mutable_data<uint8_t>() + k_cache_stride);
-    uint8_t* ptr_in = k_cache_in[i]->mutable_data<uint8_t>() - pos_diff;
+    uint8_t* ptr_in = k_cache_in[i]->mutable_data<uint8_t>() - pos;
     for (int j = 0; j < head_dim_; ++j) {
       memcpy(
         ptr_in + j * prefill_cache_len_,
```
It seems still not quite right. I need to learn your logic a bit more.

haowhsu-quic commented:
> @haowhsu-quic do you mean changes like this?
> 
> ```
> +
> +  // update v_cache
> +  std::vector<std::unique_ptr<executorch::aten::TensorImpl>>& v_cache_out =
> +      v_cache_out_[prefill_forward_name_];
> +
> +  for (int i = 0, v_cache_stride = head_dim_ * pos_diff; i < v_cache_in.size(); ++i) {
> +    v_cache_in[i]->set_data(v_cache_in[i]->mutable_data<uint8_t>() + v_cache_stride);
> +    v_cache_out[i]->set_data(v_cache_out[i]->mutable_data<uint8_t>() + v_cache_stride);
>    }
>  
>    // update k_cache
> @@ -521,7 +525,7 @@
>         i++) {
>      k_cache_in[i]->set_data(
>          k_cache_in[i]->mutable_data<uint8_t>() + k_cache_stride);
> -    uint8_t* ptr_in = k_cache_in[i]->mutable_data<uint8_t>() - pos_diff;
> +    uint8_t* ptr_in = k_cache_in[i]->mutable_data<uint8_t>() - pos;
>      for (int j = 0; j < head_dim_; ++j) {
>        memcpy(
>          ptr_in + j * prefill_cache_len_,
> ```
> 
> It seems still not quite right. I need to learn your logic a bit more.

Yes, could you update the latest change? thank you.

haowhsu-quic commented on examples/qualcomm/oss_scripts/llama/runner/io_manager.cpp:
```diff
@@ -494,10 +494,56 @@ void ShiftPointerIoMgr::prepare_prefill_io(
+  std::vector<std::unique_ptr<executorch::aten::TensorImpl>>& k_cache_in =
+      k_cache_in_[prefill_forward_name_];
+
+  size_t copied_size = pos_diff * sizeof(uint8_t);
```
`copied_size` should be `pos * sizeof(uint8_t);`

haowhsu-quic commented on examples/qualcomm/oss_scripts/llama/runner/io_manager.cpp:
```diff
@@ -494,10 +494,56 @@ void ShiftPointerIoMgr::prepare_prefill_io(
+    k_cache_in[i]->set_data(
+        k_cache_in[i]->mutable_data<uint8_t>() + k_cache_stride);
+    uint8_t* ptr_in = k_cache_in[i]->mutable_data<uint8_t>() - pos_diff;
+    for (int j = 0; j < head_dim_; ++j) {
```
Sorry, probably need to change here a bit: `for (int j = 0; j <= head_dim_; ++j) {`
I forgot we preserve extra space to prevent shifting beyond boundary.

cccclai commented:
updated with the suggestion, still incorrect. I'm trying to dump the kv cache value to confirm
```
   int64_t pos_diff = pos - last_pos_;
   std::vector<std::unique_ptr<executorch::aten::TensorImpl>>& v_cache_in =
       v_cache_in_[prefill_forward_name_];
-  for (int i = 0, v_cache_stride = head_dim_ * pos_diff; i < v_cache_in.size();
-       i++) {
-    v_cache_in[i]->set_data(
-        v_cache_in[i]->mutable_data<uint8_t>() + v_cache_stride);
+
+  // update v_cache
+  std::vector<std::unique_ptr<executorch::aten::TensorImpl>>& v_cache_out =
+      v_cache_out_[prefill_forward_name_];
+
+  for (int i = 0, v_cache_stride = head_dim_ * pos_diff; i < v_cache_in.size(); ++i) {
+    v_cache_in[i]->set_data(v_cache_in[i]->mutable_data<uint8_t>() + v_cache_stride);
+    v_cache_out[i]->set_data(v_cache_out[i]->mutable_data<uint8_t>() + v_cache_stride);
   }
 
   // update k_cache
   std::vector<std::unique_ptr<executorch::aten::TensorImpl>>& k_cache_in =
       k_cache_in_[prefill_forward_name_];
 
-  size_t copied_size = pos_diff * sizeof(uint8_t);
+  size_t copied_size = pos * sizeof(uint8_t);
 
-  for (int i = 0, k_cache_stride = pos_diff * sizeof(uint8_t); i < k_cache_in_.size();
-       i++) {
-    k_cache_in[i]->set_data(
-        k_cache_in[i]->mutable_data<uint8_t>() + k_cache_stride);
-    uint8_t* ptr_in = k_cache_in[i]->mutable_data<uint8_t>() - pos_diff;
-    for (int j = 0; j < head_dim_; ++j) {
-      memcpy(
-        ptr_in + j * prefill_cache_len_,
-        ptr_in + j * kv_cache_len_,
-        copied_size);
+  for (int i = 0; i < k_cache_in.size(); i++) {
+    uint8_t* ptr_in = k_cache_in[i]->mutable_data<uint8_t>();

+    // Move pointer forward by pos_diff
+    k_cache_in[i]->set_data(ptr_in + pos_diff);
+    // Copy data from kv_cache region to prefill_cache region for each head
+    for (int j = 0; j <= head_dim_; ++j) {
+      uint8_t* dst = ptr_in - pos + j * prefill_cache_len_;
+      const uint8_t* src = ptr_in - pos + j * kv_cache_len_;
+      memcpy(dst, src, copied_size);
     }
   }
 

```

cccclai commented:
I'll ask @kushrast to update the PR tomorrow. Probably I don't have the permission to update it...

cccclai commented:
To double check, is the k_cache shape `(head_dim + 1, seq_len - 1, num_layers)`?

haowhsu-quic commented:
I think last for loop should be:
```c++
for (int i = 0; i < k_cache_in.size(); i++) {
  // should update first to current absolute position
  k_cache_in[i]->set_data(ptr_in + pos_diff);
  uint8_t* ptr_in = k_cache_in[i]->mutable_data<uint8_t>();
  for (int j = 0; j <= head_dim_; ++j) {
    uint8_t* dst = ptr_in - pos + j * prefill_cache_len_;
    const uint8_t* src = ptr_in - pos + j * kv_cache_len_;
    memcpy(dst, src, copied_size);
  }
}
```
Shape of `k_cache_in` (single head) is `(head_dim_+1, prefill_ar_len_)` for prefill mode, `(head_dim_+1, kv_cache_len_)` for decode mode. But both of them are actually mapped to the same chunk of memory, that's why we need deep copy here. Or the data fetching would be incorrect.

haowhsu-quic commented:
Is it possible we can have the `.pte` file to help resolve issue?

cccclai commented:
> Is it possible we can have the `.pte` file to help resolve issue?

That might be tricky, because it’s an internal model. The easiest way is online debug session

cccclai commented:
Actually let me try export the stories model and verify accuracy with that. I think it will have similar issue. In the meanwhile, this is the latest 
```
void ShiftPointerIoMgr::update_kv_to_prefill_io(
  int64_t pos,
  std::vector<std::vector<executorch::aten::Tensor>>& output_tensors) {
  // update v_cache
  assert(pos <= 512);
  int64_t pos_diff = pos - last_pos_;
  std::vector<std::unique_ptr<executorch::aten::TensorImpl>>& v_cache_in =
      v_cache_in_[prefill_forward_name_];

  // update v_cache
  std::vector<std::unique_ptr<executorch::aten::TensorImpl>>& v_cache_out =
      v_cache_out_[prefill_forward_name_];

  for (int i = 0, v_cache_stride = head_dim_ * pos_diff; i < v_cache_in.size(); ++i) {
    v_cache_in[i]->set_data(v_cache_in[i]->mutable_data<uint8_t>() + v_cache_stride);
    v_cache_out[i]->set_data(v_cache_out[i]->mutable_data<uint8_t>() + v_cache_stride);
  }

  // update k_cache
  std::vector<std::unique_ptr<executorch::aten::TensorImpl>>& k_cache_in =
      k_cache_in_[prefill_forward_name_];

  size_t copied_size = pos * sizeof(uint8_t);

  for (int i = 0; i < k_cache_in.size(); i++) {
    k_cache_in[i]->set_data(ptr_in + pos_diff);
    uint8_t* ptr_in = k_cache_in[i]->mutable_data<uint8_t>();
    // Copy data from kv_cache region to prefill_cache region for each head
    for (int j = 0; j <= head_dim_; ++j) {
      uint8_t* dst = ptr_in - pos + j * prefill_cache_len_;
      const uint8_t* src = ptr_in - pos + j * kv_cache_len_;
      memcpy(dst, src, copied_size);
    }
  }

  // Setting attention mask from context_len - prefill_ar_len - i to context_len
  IO* ptr = static_cast<IO*>(data_ptr_.get());
  for (int i = prefill_ar_len_; i < pos; i++) {
    for (int j = 0; j < prefill_ar_len_; j++) {
      ptr->prefill_attention_mask[j * context_len_ + context_len_ - prefill_ar_len_ - i] = 65535;
    }
  }
}
```

cccclai commented:
Actually you can repro the accuracy with this command line with stories model
```
./qnn_llama3_2_runner --model_path hybrid_stories_qnn.pte    --tokenizer_path tokenizer.bin  --eval_mode 1 --prompt "Once" --kv_updater "ShiftPointer" --logits_scale 0.1 --output_path output.txt --num_iters 2
```
In the second iteration, It should include the previous prompt + generate output + second prompt

cccclai commented:
I'm getting the correct output from https://github.com/pytorch/executorch/pull/9662, wonder if you can take a look and see if it's correct
